Migration to Vite

palagdan opened the pull request:
Resolves #51 
Resolves #58

palagdan commented:
@blcham 

After updating the extension from .js to .jsx for all the components, I encountered an issue with loading a pipeline. Specifically, after initializing a graph, the node state is empty. I tried to investigate it using debugging, but I couldn't figure out the problem. Everything else is working fine with Vite, including static images.
![Screenshot from 2024-08-29 14-28-51](https://github.com/user-attachments/assets/aec4f26c-f5ea-46d8-ad82-2424753121e9)

palagdan commented:
# To Do After Migration to Vite

1. **Refactor Class Components to Functional Components**  
   Functional components are recommended for better performance and simplicity.

2. **Refactor `main.css`**  
   Refactor main.css into separate

3. **Fix Script Execution Issues**  
   Address any problems related to script execution.

4. **Migrate from TreeBeard to MUI Material for Script Tree Generation**  
   TreeBeard is outdated and causing warnings. Switch to MUI Material for generating the script tree due to the following warnings:

   ```bash
   TreeBeard2: Support for defaultProps will be removed from function components in a future major release. Use JavaScript default parameters instead.

   Warning: Transition2 uses the legacy childContextTypes API, which is no longer supported and will be removed in the next major release. Use React.createContext() instead.

   ...

blcham commented on public/config.js:
```diff
@@ -0,0 +1,3 @@
+window.__config__ = {
+  hello: "hello",
```
?

palagdan replied:
I tested whether I had injected this variable correctly into the index.html file. I removed it and performed a squash.

blcham commented on src/components/treebeard/ScriptsTree.jsx:
```diff
@@ -97,7 +97,7 @@ class ScriptsTree extends React.Component {
 
   render() {
-    if (this.state.data === []) {
+    if (this.state.data.length === 0) {
```
why is this better?

palagdan replied:
Since this.state.data === [] compares the references of two objects, it will always evaluate to false

palagdan replied:
Comparing arrays using === checks if the two references point to the same object, not if their contents are the same. So it always evaluates to false.

blcham replied:
ok, i wrote it because I saw that u started to assume that this.state
data is always an array and cannot be null, which is most likely fine i just wanted to check

blcham commented on vite.config.js:
```diff
@@ -1,15 +1,28 @@
+  plugins: [
+    react(),
+    envCompatible(),
+    nodePolyfills({
```
We should short comment in the file here why we added that here and link to the github comment that explains it

palagdan replied:
t is almost complete. I still need to fix a bug I mentioned earlier with the Dagre scripts. The behavior is strange because state.nodes and state.edges are updated with the correct data, but these states end up empty.

blcham replied:
? so what is the state of this? Is it commented or we do not need to do it anymore?

blcham replied:
Ok I do not see line:
`nodePolyfills({`

anymore so i guess we can resolve this comment. If so do it please.

blcham replied:
@palagdan ok

blcham reviewed (changes requested):
i assume everything is working fine now in SPipes editor, so after
comments are resolved we can merge

blcham commented on src/config/env.js:
```diff
@@ -40,3 +40,5 @@ export const getEnv = (name, defaultValue) => {
   throw new Error(`Missing environment variable: ${name}`);
 };
+
+export const APP_TITLE = getEnv("APP_TITLE", "S-Pipes UI");
```
please change to `SPipes Editor`


note, we always use only variants: 
SPipes or s-pipes but never S-Pipes

palagdan replied:
done

blcham commented on .env.development:
```diff
@@ -0,0 +1 @@
+S_PIPES_APP_TITLE=S-Pipes UI dev
```
please change to `SPipes Editor Dev`


note, we always use only variants: 
SPipes or s-pipes but never S-Pipes

palagdan replied:
done

blcham commented on .env.production:
```diff
@@ -0,0 +1 @@
+S_PIPES_APP_TITLE=S-Pipes UI
```
please change to `SPipes Editor` as this is targeted to users and not developers


note, we always use only variants: 
SPipes or s-pipes but never S-Pipes

palagdan replied:
done

palagdan commented:
@blcham 
I fixed a bug related to Dagre state management. It is a magic to me how it worked before and why it stopped working after I changed the file extension from .js to .jsx (I am sure this issue was caused by this change).

I discovered that setState() is an asynchronous function but does not return a promise, making its behavior somewhat unpredictable. The only way to ensure state updates are processed before taking action is to use a callback function as the second parameter of setState(). Unfortunately, you cannot await setState() directly.

The problem was that setState() was called multiple times within processGraph() and addNodes(), leading to an empty state. Additionally, this.renderCytoscapeElement() should be called after processGraph() has updated the nodes and edges state. However, since setState() is asynchronous, it didn't update the state in time, resulting in renderCytoscapeElement() receiving empty nodes and edges arrays. This is why the pipeline wasn’t rendered as shown in the picture I sent earlier.
```javascript

   Rest.getScript(this.state.file, this.state.transformation).then((response) => {
        this._processGraph(response);
        this.renderCytoscapeElement();
      });
```

palagdan commented:
@blcham 

I am completely stuck with this ticket. Yesterday, after fixing a bug that I described earlier, I attempted to test it in Docker (production mode). When I tried to load the scripts, the page wouldn't load due to a polyfill issue with process. The external library undici uses Node.js libraries that are not supported by the browser, which is causing this error:
![Screenshot from 2024-08-31 17-29-45](https://github.com/user-attachments/assets/3943cc04-39c5-4cd6-9d47-9c1649aa906d)
The undici library tries to access a Node.js version with the process variable:
![Screenshot from 2024-08-31 17-52-38](https://github.com/user-attachments/assets/7a58dc39-5259-4154-936d-b38fb6cae442)
Since the browser does not support Node.js libraries, process is undefined, and Vite does not support it either. For reference, see: https://github.com/vitejs/vite/issues/7384.

I attempted to use the vite-plugin-node-polyfills library, but it did not resolve the issue and instead introduced new problems, such as memory leaks during the build and strange conflicts:
```text
11.73 error during build:
11.73 [commonjs--resolver] Could not load /app/node_modules/stream-browserify/web: ENOENT: no such file or directory, open '/app/node_modules/stream-browserify/web'
11.73 file: /app/node_modules/jsonld/lib/index.js
11.73 Error: Could not load /app/node_modules/stream-browserify/web: ENOENT: no such file or directory, open '/app/node_modules/stream-browserify/web'
```
I don't know how to resolve this, as there is limited information available, and nothing seems to work.

blcham commented:
1) please rebase
2) describe e.g. package-lock file where does unidici library comes from and why it is there

palagdan commented:
this library comes from  s-forms -> jsonld -> http-client -> undici
|![Screenshot from 2024-09-19 11-58-17](https://github.com/user-attachments/assets/a51cff5e-21b7-4946-8db0-1ecb8aa60185)

blcham commented:
@palagdan 

That is very good news :) as we have in control https://gitlab.fel.cvut.cz/kbss1/jsonld-utils. 

We were thinking quite a lot if it is not good idea to remove this dependency, and include relevant content of this library manually to the s-forms. I would follow like this:

1) clone the repository
2) find out if it is possible to create new version for jsonld-utils that does not have this issue
3) if not possible we will modify s-forms to include extracted classes from jsonld-utils

palagdan commented:
@blcham

I’ve fixed the bug, and it now works correctly in production mode as well. It’s ready for merging.

blcham reviewed (approved):
LGTM

blcham commented on src/components/dagre/Dagre.jsx:
```diff
@@ -135,94 +129,88 @@ class Dagre extends React.Component {
     });
   }
 
-  //prevent session timeout
```
Why we do not need this anymore?

palagdan replied:
Fixed here https://github.com/kbss-cvut/s-pipes-editor-ui/issues/90

blcham commented on src/components/modal/ScriptExecutionModal.jsx:
```diff
@@ -96,16 +95,6 @@ class ScriptOntologyModal extends React.Component {
                       </td>
                       <td>{data[MODULE_EXECUTION_DURATION]}ms</td>
                       {/*<td><Moment unix format="DD.MM.YYYY">{data[MODULE_EXECUTION_START_DATE]/1000}</Moment></td>*/}
-                      <td>
```
Do we still show execution start and finish date?

blcham replied:
BTW, execution is not working which you know i assume as we targeted it in #62

palagdan replied:
Fixed here https://github.com/kbss-cvut/s-pipes-editor-ui/issues/95

blcham commented:
@palagdan We still need to answer the comments above which I included into the first bullet of #62

blcham commented on src/components/dagre/Dagre.jsx:
```diff
@@ -58,10 +57,6 @@ const rankDirOptions = [
   { text: "TopBottom", key: "TB", value: "TB" },
 ];
 
-const websocketURL = new URL("/rest/notifications", window.location.href);
```
This does not work; please make a new issue for that. 
To reproduce: 
- open a script
- modify the script using a textual editor outside of SPipes editor

blcham replied:
@palagdan

palagdan replied:
Fixed here https://github.com/kbss-cvut/s-pipes-editor-ui/issues/90
